core/domain/stats_jobs_one_off.py

```diff
@@ -1275,3 +1275,59 @@ def reduce(key, stringified_values):
                     'all: %s sum:%s' % (
                         key, state_name, all_state_hit[state_name],
                         sum_state_hit[state_name]),)
+
+
+class RegenerateMissingStatsModels(jobs.BaseMapReduceOneOffJobManager):
+    """A one-off job to regenerate stats models which were missing due to
+    incorrect handling of exploration reverts. If a model is missing at version
+    x, we will regenerate all models from version x-1 till the max version of
+    the exploration.
+    """
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [exp_models.ExplorationModel]
+
+    @staticmethod
+    def map(exp):
+        all_models = (
+            stats_models.ExplorationStatsModel.get_multi_stats_models(
+                [exp_domain.ExpVersionReference(exp.id, version)
+                 for version in range(1, exp.version + 1, 1)]))
+        first_missing_version = -1
+        for i, model in enumerate(all_models):
+            if model is None:
+                first_missing_version = i + 1
+
+        if first_missing_version == -1:
+            yield ('No change', exp.id)
+
+        for i in range(first_missing_version - 1, exp.version + 1, 1):
+            commit_log = exp_models.ExplorationCommitLogEntryModel.get_commit(
+                exp.id, i)
+            exp_at_version_i = exp_models.ExplorationModel.get_version(
+                exp.id, i)
+            # Delete all old models.
+            if all_models[i - 1] is None:
+                if commit_log.commit_type != 'revert':
+                    yield ('Missing model without revert commit', exp.id)
```

Comment:
This should ideally not be hit. But added it anyway as we aren't sure that reverts are the only actions causing the issue.

Reply:
Hmm, could an audit confirm this for us?

Reply:
So I wanted to do this job that regenerates models as it has been a serious issue which has been logging 400k-500k errors every release. And as this is a costly job (which runs through every exp version), I thought of clubbing the regeneration for cases we know will happen and get info on if something else is missing too. I think I should probably add the exp_version in the yield too so that we can investigate it better. 

But does this sound ok? (Basically this job fixes known issues, and audits to see if any unknown issues exist).

Reply:
Good idea to audit for unknown issues and yield extra info that's useful. Agree with adding extra info to ease debugging. Overall, I agree re the general approach.

Reply:
SGTM